File: Services/src/CodeRunner/CodeRunner.Model/CompileError.cs
using System.Collections.Generic;

namespace CSC.CodeRunner.Model
{
	/// <summary>
	/// An error encountered during compilation.
	/// </summary>
	public class CompileError
	{
		/// <summary>
		/// The line number of the error.
		/// </summary>
		public int LineNumber { get; set; }

		/// <summary>
		/// The column number of the error.
		/// </summary>
		public int ColumnNumber { get; set; }

		/// <summary>
		/// The error message.
		/// </summary>
		public string Message { get; set; }

		/// <summary>
		/// The full error message, including context.
		/// </summary>
		public string FullError { get; set; }
	}
}

File: Runners/JavaCodeRunner/src/main/java/csc/coderunner/CompileError.java
package csc.coderunner;

import javax.tools.Diagnostic;
import javax.tools.JavaFileObject;

/**
 * An error encountered during compilation.
 */
public class CompileError
{
	private int lineNumber;
	private int columnNumber;
	private String message;
	private String fullError;
	
	/**
	 * Constructor.
	 * @param diagnostic The compiler diagnostic.
	 * @param lineNumberOffset The offset to add to the line number.
	 */
	public CompileError(Diagnostic<? extends JavaFileObject> diagnostic, int lineNumberOffset)
	{
		this.lineNumber = (int)(diagnostic.getLineNumber() + lineNumberOffset);
		this.columnNumber = (int)diagnostic.getColumnNumber();
		this.message = diagnostic.getMessage(null);
		
		String rawError = diagnostic.toString();
		String errorPrefix = "error: ";
		int rawErrorBegin = rawError.indexOf(errorPrefix);
		if (rawErrorBegin != -1)
		{
			fullError = String.format("Error on line %d: %s",
				lineNumber,
				rawError.substring(rawErrorBegin + errorPrefix.length()));
		}
		else
		{
			fullError = rawError;
		}
	}
	
	/**
	 * @return The line number of the error.
	 */
	public long getLineNumber()
	{
		return lineNumber;
	}

	/**
	 * @return The column number of the error.
	 */
	public long getColumnNumber()
	{
		return columnNumber;
	}

	/**
	 * @return The error message.
	 */
	public String getMessage()
	{
		return message;
	}

	/**
	 * @return The full error message, including context.
	 */
	public String getFullError()
	{
		return fullError;
	}
}
